osu.Game/Overlays/SocialOverlay.cs

```diff
@@ -62,62 +57,149 @@ public SocialOverlay()
-                        AutoSizeAxes = Axes.Y,
-                        Margin = new MarginPadding { Top = 20 },
-                        Spacing = new Vector2(10f),
-                    }
-                },
+                    Users = null;
+                    queryChangedDebounce = Scheduler.AddDelayed(updateSearch, 500);
+                }
             };
 
-            Add(loading = new LoadingAnimation());
+            currentQuery.BindTo(Filter.Search.Current);
+
+            Filter.Tabs.Current.ValueChanged += sortCriteria => Scheduler.AddOnce(updateSearch);
+
+            Scheduler.AddOnce(updateSearch); // so it displays something once it's first opened
         }
 
         [BackgroundDependencyLoader]
         private void load(APIAccess api)
         {
-            if (Users == null)
-                reloadUsers(api);
+            this.api = api;
         }
 
-        private void reloadUsers(APIAccess api)
+        private void recreatePanels(PanelDisplayStyle displayStyle)
         {
-            Users = null;
+            if (panels != null)
+            {
+                panels.FadeOut(200);
+                panels.Expire();
+                panels = null;
+            }
+
+            if (Users == null)
+                return;
 
-            // no this is not the correct data source, but it's something.
-            var request = new GetUsersRequest();
-            request.Success += res =>
+            var newPanels = new FillFlowContainer<UserPanel>
             {
-                Users = res.Select(e => e.User);
-                loading.Hide();
+                RelativeSizeAxes = Axes.X,
+                AutoSizeAxes = Axes.Y,
+                Spacing = new Vector2(10f),
+                Margin = new MarginPadding { Top = 10 },
+                ChildrenEnumerable = Users.Select(u =>
+                {
+                    UserPanel panel = new UserPanel(u)
+                    {
+                        Anchor = Anchor.TopCentre,
+                        Origin = Anchor.TopCentre
+                    };
+                    switch (displayStyle)
+                    {
+                        case PanelDisplayStyle.Grid:
+                            panel.Width = 300;
+                            break;
+                        default:
+                            panel.RelativeSizeAxes = Axes.X;
+                            break;
+                    }
+                    panel.Status.BindTo(u.Status);
+                    return panel;
+                })
             };
 
-            api.Queue(request);
-            loading.Show();
+            LoadComponentAsync(newPanels, p =>
+            {
+                if (panels != null)
+                    ScrollFlow.Remove(panels);
+
+                ScrollFlow.Add(panels = newPanels);
+            });
         }
 
-        public void APIStateChanged(APIAccess api, APIState state)
+        private GetUsersRequest getUsersRequest;

```

Comment:
i don't see any reason to store both of these requests separately – it just increases complexity. we should just store the currently running request for the currently displayed tab.

Reply:
I tried with having just an `APIRequest` but as only `APIRequest<T>` has the Success event it won't work. Also both Requests return data in a different structure.
I could try merging those two request classes and just have the type fetched as an enum just like the other criterias would be once that comes. `new GetUsersRequest(UserType.Friends)`
UserType could include `Friends, Other, Team` , etc.
Does that sound better?